data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/RenameKeyProcessor.java

```diff
@@ -59,14 +69,27 @@ public Collection<Record<Event>> doExecute(final Collection<Record<Event>> recor
                         continue;
                     }
 
-                    if (entry.getFromKey().equals(entry.getToKey()) || !recordEvent.containsKey(entry.getFromKey())) {
+                    if (Objects.nonNull(entry.getFromKey()) && (entry.getFromKey().equals(entry.getToKey()) || !recordEvent.containsKey(entry.getFromKey()))) {
                         continue;
                     }
-
                     if (!recordEvent.containsKey(entry.getToKey()) || entry.getOverwriteIfToKeyExists()) {
-                        final Object source = recordEvent.get(entry.getFromKey(), Object.class);
-                        recordEvent.put(entry.getToKey(), source);
-                        recordEvent.delete(entry.getFromKey());
+                        if(Objects.nonNull(entry.getFromKey())) {
+                            final Object source = recordEvent.get(entry.getFromKey(), Object.class);
+                            recordEvent.put(entry.getToKey(), source);
+                            recordEvent.delete(entry.getFromKey());
+                        }
+                        if(Objects.nonNull(entry.getFromKeyPattern())) {
+                            Map<String,Object> eventMap = recordEvent.toMap();
+                            Pattern pattern = Pattern.compile(entry.getFromKeyPattern());
+                            for (Map.Entry<String, Object> eventEntry : eventMap.entrySet()) {
+                                final String key = eventEntry.getKey();
+                                final Object value = eventEntry.getValue();
+                                if (pattern.matcher(key).matches()) {
```

Comment:
Our current behavior for most processors is not to override keys when it already exists. We should follow that pattern only only rename on the first matched key. So, if `overwriteIfToKeyExists` is false, break from the for-loop after this condition. If `overwriteIfToKeyExists` is true, then overwrite it each time with the behavior you have now.

We could possibly have this have a configurable behavior to let users customize the behavior when multiple keys match. But, for now, I think a simple behavior like this works.

Reply:
Added the fix.